main.cpp

```diff
@@ -454,7 +481,84 @@ void Mainloop::loop()
                 (*e)->print_statistics();
             }
         }
+
+        _timeout_process_del(false);
     }
+
+    _timeout_process_del(true);
+}
+
+void Mainloop::_timeout_process_del(bool del_all)
+{
+    for (uint8_t i = 0; i < MAX_TIMEOUT; i++) {
+        Timeout *t = _timeout_list[i];
+
+        if (t == NULL) {
+            continue;
+        }
+        if (!del_all && !t->remove_me) {
+            continue;
+        }
+
+        _timeout_list[i] = NULL;
+        remove_fd(t->fd);
+        delete t;
+    }
+}
+
+Timeout *Mainloop::timeout_add(uint32_t timeout_ms, bool (*cb)(void *data), const void *data)
+{
+    struct itimerspec ts;
+    Timeout *t = new Timeout();
+    bool placed = false;
+    uint8_t index_placed;
+
+    null_check(t, NULL);
+
+    for (index_placed = 0; index_placed < MAX_TIMEOUT; index_placed++) {
+        if (_timeout_list[index_placed] == NULL) {
+            _timeout_list[index_placed] = t;
+            placed = true;
+            break;
+        }
```

Comment:
I'm using array because you said that we should only support a few timerfd's if we needed more we would need to implement a sorted list of timeouts, but even if you want to keep this limit I don't see any problems in moving to std::list.